### PR TITLE
check max_changed_partitions at construction-time

### LINKED ISSUE #1908: Polars privacy map fail but measurement goes through
**Bug Description**

We construct a polars measurement on a dataset with two columns: "region" and "income". The polars plan computes the average income by region and adds noise with an arbitrary scale of one.

The lazyframe domain we use has a margin for the "region", but only includes `max_partition_length` and omits `max_num_partitions`.

BUG: The measurement can be constructed and executed on data but the privacy map fails with `OpenDPException: 
  FailedFunction("max_num_partitions must be known when the metric is not sensitive to ordering (SymmetricDistance)")`

Since the privacy map fails, we expected the measurement to also fail. Is this a bug or is the current behaviour to be expected?

**To Reproduce**

````python
import opendp.prelude as dp
import polars as pl

dp.enable_features("contrib")

# Minimal dataframe
NB_REGIONS = 3
NB_ROWS = 6

df = pl.DataFrame({"region": [1, 1, 2, 2, 3, 3], "income": [1000.0, 1000.0, 2000.0, 2000.0, 3000.0, 3000.0]}, schema={"region": pl.Int64, "income": pl.Float64})
lf = df.lazy()

# Minimal domain
lf_domain = dp.lazyframe_domain([
    dp.series_domain("region", dp.atom_domain(T=dp.i64)),
    dp.series_domain("income", dp.atom_domain(T=dp.f64))
])
lf_domain = dp.with_margin(lf_domain, by=[], public_info="lengths", max_partition_length=NB_ROWS)

# Group-by query: average income per region, add noïse with scale of 1.0
income_lower_bound, income_upper_bound = 1_000, 100_000
plan = lf.group_by("region").agg([
    pl.col("income").dp.mean(bounds=(income_lower_bound, income_upper_bound), scale=1.0)
]).sort("income")

# Bad domain: add a margin that does not include max_num_partitions
bad_domain = dp.with_margin(lf_domain, by=["region"], public_info="lengths", max_partition_length=NB_ROWS)

# Construct and collect a measurement with bad_domain works
bad_meas = dp.make_private_lazyframe(bad_domain, dp.symmetric_distance(), dp.max_divergence(T=float), plan)
bad_res = bad_meas(lf).collect()

# Privacy map fails
bad_meas.map(1)
```

### PR DESCRIPTION
**Pull request by Shoeboxam**
Close #1908

### CODE REVIEW

**Comment by Shoeboxam** [comment]
* **#2059** <a href="https://app.graphite.dev/github/pr/opendp/opendp/2059?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @Shoeboxam and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: python/test/test_polars.py [hunk 1]
```diff
@@ -821,3 +821,23 @@ def test_to_physical_unordered():
...
+        dp.series_domain("region", dp.atom_domain(T=dp.i64)),
+        dp.series_domain("income", dp.atom_domain(T=dp.f64))
+    ])
+    lf_domain = dp.with_margin(lf_domain, by=["region"], public_info="lengths", max_partition_length=6)
+
+    from opendp.domains import _lazyframe_from_domain
+    lf = _lazyframe_from_domain(lf_domain)
+
+    # sum of income per region, add noise with scale of 1.0
+    plan = lf.group_by("region").agg([
```
Comment (mccalluc) [inline python/test/test_polars.py:837]: Slight preference for keeping the `importorskip` as late as possible, so the first part of the test runs whether or not polars is installed. If there's an indirect dependency on polars, that could be handled with `optional_dependency`.

```suggestion
    lf_domain = dp.lazyframe_domain([
        dp.series_domain("region", dp.atom_domain(T=dp.i64)),
        dp.series_domain("income", dp.atom_domain(T=dp.f64))
    ])
    lf_domain = dp.with_margin(lf_domain, by=["region"], public_info="lengths", max_partition_length=6)

    from opendp.domains import _lazyframe_from_domain
    lf = _lazyframe_from_domain(lf_domain)

    # sum of income per region, add noise with scale of 1.0
    pl = pytest.importorskip('polars')
    plan = lf.group_by("region").agg([
```

File: python/test/test_polars.py [hunk 2]
```diff
@@ -821,3 +821,23 @@ def test_to_physical_unordered():
...
+        pl.col("income").dp.sum(bounds=(1_000, 100_000), scale=1.0)
+    ])
+
+    # since there are an unknown number of partitions, and each partition has non-zero sensitivity, sensitivity is undefined
+    with pytest.raises(dp.OpenDPException, match='.*max_num_partitions must be known when the metric is not sensitive to ordering.*'):
```
Comment (mccalluc) [inline python/test/test_polars.py:842]:
```suggestion
    with pytest.raises(dp.OpenDPException, match='max_num_partitions must be known when the metric is not sensitive to ordering'):
```
regex can match substring.

File: rust/src/transformations/make_stable_expr/expr_sum/mod.rs [hunk 3]
```diff
@@ -125,6 +125,11 @@ where
         _ => TI::Sum::inf_cast(*d_in)?.inf_mul(&l.alerting_abs()?.total_max(u)?),
     };
 
+    // 'mnp_check: this invariant is used later
```
Comment (mccalluc) [inline rust/src/transformations/make_stable_expr/expr_sum/mod.rs:128]: I'm expecting to see `'mnp_check` in the code here, but it's just in an error message. Is that right?

**Review by mccalluc — COMMENTED** [review]
Some little comments, but the biggest headache seems to be hashbrown: Whatever the fix for that is, it might belong in its own PR that others can be based on.

File: rust/src/transformations/make_stable_expr/expr_sum/mod.rs (hunk 3 above)
Comment (Shoeboxam) [reply rust/src/transformations/make_stable_expr/expr_sum/mod.rs:128]: yep, the error message should never be seen by the user, and it references another location by a comment.

**Review by mccalluc — APPROVED** [review]
Thanks!

**Pull request merged**